### ISSUE TITLE
Dataset normalization

### ISSUE THREAD

**Issue by TimurIbrayev** [MadryLab/cifar10_challenge#15]
Hello,

I am trying to re-implement your CIFAR10 adv. training in PyTorch and maybe some of the questions will be based on my limited knowledge of TensorFlow.

I have couple of questions regarding CIFAR10 dataset normalization. In PyTorch, the entire dataset is usually normalized as the dataset is loaded through the loader by adding normalization as one of transformations (after converting image to tensor to be in range [0,1]). Also, the normalization is usually implemented by specifying per-channel mean and stddev computed for the **entire** dataset. Hence, my questions are the following:

1. What is the reason you are implementing "per_image_standardization" as part of the model rather than normalization over entire dataset as preprocessing? Is it to keep original samples in the range of 0-255 and to perform perturbations in that range?

2. Is there any difference between implementing standardization/normalization using per-channel mean and stddev computed for entire dataset (case of PyTorch) **and** mean and stddev computed for each separate image (case of tf.image.per_image_standardization)? As far as I can tell the end aim is essentially the same for both cases: to have samples with zero mean and unit variance. But I think in case of TensorFlow, as the sample will be perturbed, normalization will change correspondingly to keep the input distribution to the model consistently with zero mean and unit variance.

Thank you and sorry for verbosity: wanted to make sure I delivered my concerns properly.

**Comment by dtsip** [comment]
Hi,

1. Since we define an epsilon constraint in input space (each pixel can change by at most 8/255 by the adversary) we need to ensure that this constraint also applies in the case of preprocessing. In general, there are two options here: either perform normalization during preprocessing and *also normalize eps* or include normalization as part of the model in which case the adversary will just backprop through it and everything will work fine. We chose the second option for implementation convenience.

2. There was no particular reason for this choice. We just followed the CIFAR10 pipeline we found in the TF tutorial section at the time. Normalizing in a different way (or not at all) should produce qualitatively similar results.

In general, we recently released a library for adversarial training in PyTorch https://github.com/MadryLab/robustness that might be a useful reference.
